solidity/contracts/isms/ccip-read/AbstractCcipReadIsm.sol

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.8.0;
+
+// ============ Internal Imports ============
+import {IInterchainSecurityModule} from "../../interfaces/IInterchainSecurityModule.sol";
+import {ICcipReadIsm} from "../../interfaces/isms/ICcipReadIsm.sol";
+import {IMailbox} from "../../interfaces/IMailbox.sol";
+import {Message} from "../../libs/Message.sol";
+import {AbstractMultisigIsm} from "../multisig/AbstractMultisigIsm.sol";
+
+/**
+ * @title AbstractCcipReadIsm
+ * @notice An ISM that allows arbitrary payloads to be submitted and verified on chain
+ * @dev https://eips.ethereum.org/EIPS/eip-3668
+ * @dev The AbstractCcipReadIsm provided by Hyperlane is left intentially minimalist as
+ * the range of applications that could be supported by a CcipReadIsm are so broad. However
+ * there are few things to note when building a custom CcipReadIsm.
+ *
+ * 1. `getOffchainVerifyInfo` should revert with a `OffchainLookup` error, which encodes
+ *    the data necessary to query for offchain information
+ * 2. For full CCIP Read specification compatibility, CcipReadIsm's should expose a function
+ *    that in turn calls `process` on the configured Mailbox with the provided metadata and
+ *    message. This functions selector should be provided as the `callbackFunction` payload
+ *    for the OffchainLookup error
```

Comment:
IIUC we wouldn't need to do this "pointer indirection" because the `Mailbox.process` would bubble up the revert reason from the `ism.verify` call

Reply:
Let me be a little pedantic, spec wise, for a second. There's two ways for CCIP Read contracts to work,

1. Call some function, ie. `getOffchainVerifyInfo` which describes how to get and submit the off chain info to a callback
2. Call the function you actually want to call, ie. `Mailbox.process`, which reverts in the absence of off chain data but describes how to get the data. Allowing the caller to then call the function again.

I opted for option 1 because the Mailbox is not CCIP Read aware, so we break a core invariant in CCIP Read that states `If the sender field does not match the address of the contract that was called, return an error to the caller and stop.`. The Mailbox would need to catch the ISM error and revert with its own `OffchainLookup` error in turn.

The Relayer doesn't respect the callback function anyway so it's kind of irrelevant. But IMO option 1 at least has the option of being totally spec compliant if we want

Reply:
that checks out
I do still prefer option 2 in theory but requiring the Mailbox to be CCIP read aware is quite a departure from the expected scope